sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java

```diff
@@ -49,108 +54,202 @@ public HttpLoggingPolicy(HttpLogOptions httpLogOptions) {
-                        });
-                    } catch (Exception e) {
-                        reqBodyLoggingMono = Mono.error(e);
-                    }
+                String contentType = request.getHeaders().getValue("Content-Type");
+                long contentLength = getContentLength(logger, request.getHeaders());
+
+                if (shouldBodyBeLogged(contentType, contentLength)) {
+                    requestLoggingMono = FluxUtil.collectBytesInByteBufferStream(request.getBody()).flatMap(bytes -> {
+                        requestLogMessage.append(contentLength)
+                            .append("-byte body:")
+                            .append(System.lineSeparator())
+                            .append(prettyPrintIfNeeded(logger, contentType, new String(bytes, StandardCharsets.UTF_8)))
+                            .append(System.lineSeparator())
+                            .append("--> END ")
+                            .append(request.getHttpMethod())
+                            .append(System.lineSeparator());
+
+                        return Mono.just(requestLogMessage.toString());
+                    });
                 } else {
-                    logger.info("{}-byte body: (content not logged)", contentLength);
-                    logger.info("--> END {}", request.getHttpMethod());
+                    requestLogMessage.append(contentLength)
+                        .append("-byte body: (content not logged)")
+                        .append(System.lineSeparator())
+                        .append("--> END ")
+                        .append(request.getHttpMethod())
+                        .append(System.lineSeparator());
                 }
             }
         }
-        return reqBodyLoggingMono;
+
+        return requestLoggingMono.doOnNext(logger::info);
     }
 
-    private void formatAllowableHeaders(Set<String> allowedHeaderNames, HttpHeaders requestResponseHeaders,
-        ClientLogger logger) {
-        Set<String> lowerCasedAllowedHeaderNames = allowedHeaderNames.stream().map(String::toLowerCase)
-            .collect(Collectors.toSet());
-        StringBuilder sb = new StringBuilder();
-        for (HttpHeader header : requestResponseHeaders) {
-            String headerName = header.getName();
-            sb.append(headerName).append(":");
-            if (lowerCasedAllowedHeaderNames.contains(headerName.toLowerCase(Locale.ROOT))) {
-                sb.append(header.getValue());
+    /*
+     * Logs thr HTTP response.
+     *
+     * @param logger Logger used to log the response.
+     * @param response HTTP response returned from Azure.
+     * @param startNs Nanosecond representation of when the request was sent.
+     * @return A Mono containing the HTTP response.
+     */
+    private Mono<HttpResponse> logResponse(final ClientLogger logger, final HttpResponse response, long startNs) {
+        int numericLogLevel = LoggingUtil.getEnvironmentLoggingLevel().toNumeric();
```

Comment:
This line can be moved inside `shouldLoggingBeSkipped()` as well and the method can be parameter-less.

Reply:
`numericLogLevel` is a parameter that is passed into adding headers to the log message, this is done to maintain a consistent logging state while the message is being generated. This will stay for now.